### PR TITLE
Implement minimal validation of sls formula

### PR DESCRIPTION
**Pull request by MalloZup**
ready to review

### CODE REVIEW

**Comment by MalloZup** [comment]
It is same as the other one but we could mimic same behavior. We might put the data for pillar in example

**Comment by MalloZup** [comment]
looks good to be merged imho @ayoub-belarbi cc

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Here some few comments:

1. I wouldn't create the example folder, more not used files... Could we use the existing ones in test, or create a file (the content is simple) on the fly (like you do with grains)?

2. May options are not covered if we use th current pillar, we should change it.
  1. One use case for create
  2. One use case for join
  3. One use case for remove
3. The cluster configure entry is not covered.

Find here a more complete pillar file:
https://github.com/SUSE/ha-sap-terraform-deployments/blob/master/pillar_examples/libvirt/performance_optimized/cluster.sls

**Comment by MalloZup** [comment]
@arbulu89 thx for review. I have adressed all your point , take a look once you have time.

TIA :palm_tree:

**Review by arbulu89 — CHANGES_REQUESTED** [review]
I think I didn't explain correctly about the methods.
I don't talk about the `method` in configure.

This is how the cluster works.
1. If the hostname of the machine running the formula matches with the entry of `init` it will run the cluster initialization
2. If the hostname of the machine running the formula is in the `remove` list, it will remove the current node from the cluster
3. Otherwise, the machine will join the cluster trying to connect to the node with name of the `init`/

This is what we need to test

**Comment by MalloZup** [comment]
ok like this https://github.com/SUSE/habootstrap-formula/blob/25b489849344bf631110cb86dd4ccbff1066208d/test/pillar/cluster.sls#L2

File: ci/validate-formula.sh [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  '*':
+    - cluster
+EOF
+
+echo "top.sls and pillar --> DONE"
+echo
+
+
+echo "==========================================="
+echo " Using primary host                        "
```
Comment (arbulu89) [inline ci/validate-formula.sh]: I would change this to:
` Using primary host - Running init`

File: ci/validate-formula.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+echo "==========================================="
+
+cat >minion <<EOF
+root_dir: $PWD
+id: travis
+EOF
+
```
Comment (arbulu89) [inline ci/validate-formula.sh:70]: I would add here the grains file to make it easier to understand:
```
cat >grains <<EOF
host: hana01
EOF
```

File: ci/validate-formula.sh [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+sudo salt-call state.show_highstate --local --file-root=./ --config-dir=. --pillar-root=test/pillar  --retcode-passthrough -l debug
+
+echo 
+echo "==========================================="
+echo " Using secondary host                      "
```
Comment (arbulu89) [inline ci/validate-formula.sh]: Change to `Using secondary host - Running join

File: ci/validate-formula.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+echo "==========================================="
+
+generate_cluster_pillar "init: 'hana02'"
```
Comment (arbulu89) [inline ci/validate-formula.sh]: Change this to `hana01`

File: ci/validate-formula.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+set -e
+
+## this will generate the pillar
+function generate_cluster_pillar {
+  METHOD="$1"
+  cat > test/pillar/cluster.sls <<EOF
+  cluster:
+    name: 'hacluster'
+    $METHOD
+    remove: hana02
```
Comment (arbulu89) [inline ci/validate-formula.sh]: I would update this to `hana03` it makes easier to understand the concept in my opinion.
Actually, it should be:
```
remove: ['hana03']
```

**Review by arbulu89 — CHANGES_REQUESTED** [review]
With the current parameters it would work, but I see it quite complex to follow the path.
I suggest some changes to make it easier

File: ci/validate-formula.sh (hunk 5 above)
Comment (MalloZup) [reply ci/validate-formula.sh]: ok

File: ci/validate-formula.sh (hunk 4 above)
Comment (MalloZup) [reply ci/validate-formula.sh]: done

File: ci/validate-formula.sh (hunk 2 above)
Comment (MalloZup) [reply ci/validate-formula.sh:70]: done thx

File: ci/validate-formula.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+echo
+
+
+echo "==========================================="
+echo "Using primary host - Running init"
+echo "==========================================="
+
+cat >minion <<EOF
+root_dir: $PWD
+host: hana01
```
Comment (arbulu89) [inline ci/validate-formula.sh]: This has to go in `grains` file, otherwise it's not working.

File: ci/validate-formula.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+set -e
+
+## this will generate the pillar
+function generate_cluster_pillar {
+  METHOD="$1"
+  cat > test/pillar/cluster.sls <<EOF
+  cluster:
+    name: 'hacluster'
+    $METHOD
+    remove: hana03
```
Comment (arbulu89) [inline ci/validate-formula.sh]: This has to be (remove is a list):
```
remove: ['hana03']
```

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Still some fixes needed

**Review by arbulu89 — APPROVED** [review]

**Pull request merged**